File: src/test/java/projects/go/coffee_machine/stage5/main.go
package main

import "fmt"

func main() {
	var (
		money  = 550
		water  = 400
		milk   = 540
		beans  = 120
		cups   = 9
		choice int
	)

	printState := func() {
		fmt.Println("The coffee machine has:")
		fmt.Printf(
			"%d of water\n%d of milk\n%d of coffee beans\n%d of disposable cups\n$%d of money\n",
			water, milk, beans, cups, money,
		)
	}

	takeMoney := func() {
		fmt.Printf("I gave you $%d\n", money)
		money = 0
	}

	sellCoffee := func() {
		fmt.Print("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino: ")
		fmt.Scanf("%d\n", &choice)

		makeCoffee := func(getMoney, needWater, needMilk, needBeans int) {
			if water < needWater {
				fmt.Println("Sorry, not enough water!")
			} else if milk < needMilk {
				fmt.Println("Sorry, not enough milk!")
			} else if beans < needBeans {
				fmt.Println("Sorry, not enough beans!")
			} else if cups < 1 {
				fmt.Println("Sorry, not enough cups!")
			} else {
				fmt.Println("I have enough resources, making you a coffee!")
				money += getMoney
				water -= needWater
				milk -= needMilk
				beans -= needBeans
				cups--
			}
		}

		switch choice {
		case 1:
			makeCoffee(4, 250, 0, 16)
		case 2:
			makeCoffee(7, 350, 75, 20)
		case 3:
			makeCoffee(6, 200, 100, 12)
		}
	}

	fillMachine := func() {
		var next int
		fmt.Print("Write how many ml of water do you want to add: ")
		fmt.Scanf("%d\n", &next)
		water += next
		fmt.Print("Write how many ml of milk do you want to add: ")
		fmt.Scanf("%d\n", &next)
		milk += next
		fmt.Print("Write how many grams of coffee beans do you want to add: ")
		fmt.Scanf("%d\n", &next)
		beans += next
		fmt.Print("Write how many disposable cups of coffee do you want to add: ")
		fmt.Scanf("%d\n", &next)
		cups += next
	}

Loop:
	for {
		var action string
		fmt.Print("\nWrite action (buy, fill, take, remaining, exit): ")
		fmt.Scanf("%s\n", &action)
		switch action {
		case "buy":
			sellCoffee()
		case "fill":
			fillMachine()
		case "take":
			takeMoney()
		case "remaining":
			printState()
		case "exit":
			break Loop
		}
	}
}

File: src/test/java/projects/javascript/coffee_machine/stage5/main.js
const input = require('sync-input');

var money  = 550;
var water  = 400;
var milk   = 540;
var beans  = 120;
var cups   = 9;

function printState() {
	console.log("The coffee machine has:")
	console.log(
		water + " of water \n" +
		milk + " of milk \n" +
		beans + " of coffee beans \n" +
		cups + " of disposable cups \n" +
		money + " of money")
}

function takeMoney() {
	console.log("I gave you $" + money)
	money = 0
}

function sellCoffee() {
	let choice = input("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino: ");

	function makeCoffee(getMoney, needWater, needMilk, needBeans) {
		if (water < needWater) {
			console.log("Sorry, not enough water!")
		} else if (milk < needMilk) {
			console.log("Sorry, not enough milk!")
		} else if (beans < needBeans) {
			console.log("Sorry, not enough beans!")
		} else if (cups < 1) {
			console.log("Sorry, not enough cups!")
		} else {
			console.log("I have enough resources, making you a coffee!")
			money += getMoney
			water -= needWater
			milk -= needMilk
			beans -= needBeans
			cups--
		}
	}

	if (choice === '1') {
		makeCoffee(4, 250, 0, 16)
	} else if (choice === '2') {
		makeCoffee(7, 350, 75, 20)
	} else if (choice === '3') {
		makeCoffee(6, 200, 100, 12)
	}
}

function fillMachine() {
	water += Number(input("Write how many ml of water do you want to add: "))
	milk += Number(input("Write how many ml of milk do you want to add: "))
	beans += Number(input("Write how many grams of coffee beans do you want to add: "))
	cups += Number(input("Write how many disposable cups of coffee do you want to add: "))
}


while (true) {
	action = input("Write action (buy, fill, take, remaining, exit): ")

	if (action === 'buy') {
		sellCoffee()
	} else if (action === 'fill') {
		fillMachine()
	} else if (action === "take") {
		takeMoney()
	} else if (action === "remaining") {
		printState();
	} else if (action === 'exit') {
		break
	}
}
